wgpu-hal/src/metal/adapter.rs

```diff
@@ -560,28 +490,82 @@ const LAYERED_RENDERING_SUPPORT: &[MTLFeatureSet] = &[
 
 const FUNCTION_SPECIALIZATION_SUPPORT: &[MTLFeatureSet] = &[
     MTLFeatureSet::iOS_GPUFamily1_v3,
-    MTLFeatureSet::iOS_GPUFamily2_v3,
-    MTLFeatureSet::iOS_GPUFamily3_v2,
-    MTLFeatureSet::iOS_GPUFamily4_v1,
-    MTLFeatureSet::iOS_GPUFamily5_v1,
     MTLFeatureSet::tvOS_GPUFamily1_v2,
     MTLFeatureSet::macOS_GPUFamily1_v2,
-    MTLFeatureSet::macOS_GPUFamily2_v1,
 ];
 
 const DEPTH_CLIP_MODE: &[MTLFeatureSet] = &[
     MTLFeatureSet::iOS_GPUFamily4_v1,
-    MTLFeatureSet::iOS_GPUFamily5_v1,
     MTLFeatureSet::tvOS_GPUFamily1_v3,
     MTLFeatureSet::macOS_GPUFamily1_v1,
-    MTLFeatureSet::macOS_GPUFamily2_v1,
 ];
 
-impl super::PrivateCapabilities {
-    fn version_at_least(major: u32, minor: u32, needed_major: u32, needed_minor: u32) -> bool {
-        major > needed_major || (major == needed_major && minor >= needed_minor)
+#[repr(C)]
+#[derive(Clone, Copy, Debug)]
+#[allow(clippy::upper_case_acronyms)]
+struct NSOperatingSystemVersion {
+    major: usize,
+    minor: usize,
+    patch: usize,
+}
+
+const OS_NOT_SUPPORT: (usize, usize) = (10000, 0);
+// Metal was first introduced on iOS 8.0
+const IOS_MIN_VERSION: (usize, usize) = (8, 0);
+// Metal was first introduced on macOS 10.11
+const MAC_MIN_VERSION: (usize, usize) = (10, 11);
+
+impl NSOperatingSystemVersion {
+    fn at_least(
```

Comment:
this seems a bit strange to me. Shouldn't `self` already know if it's for mac or not? Why are we providing more information about `self` in an argument?